### DISCUSSION TITLE
A simple "hello world" script takes 84MB RAM -- as designed?

### DISCUSSION THREAD

**Discussion by jhuckaby** [bpftrace/bpftrace#2892] [Q&A] [answered]
Hello, I have a very simple bpftrace script:

```
#!/usr/bin/bpftrace

interval:s:1
{
	printf("Hello.\n");
}
```

Running this shows that the bpftrace process is using about 84MB of resident memory, via top:

```
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
25683 root      20   0  287620  84040  70464 S   0.0  8.7   0:00.04 bpftrace
```

This is on Amazon Linux 2, but I am seeing similar numbers on Ubuntu 23 and CentOS 9.

I am just wondering if this is as designed, and expected?  It uses even more (like 130MB) for the included `tcplife.bt` script.

Just curious if this is how it be like it is.

**Comment by danobi** [comment, marked as answer]
Yeah, it's currently pretty heavy cuz of LLVM as well as all the other deps we pull in. I expect this to get much better when we land AOT compilation. The final binary there will only contain runtime components and BPF bytecode. I've recently picked up that project again. Some more details here https://dxuuu.xyz/aot-bpftrace.html

**Reply by jhuckaby** [reply]
Oh this looks awesome, thank you!  This also answers my other question about creating compiled btftrace binaries.  I can't wait for AOT!  Thank you for your work on it!  🙏🏻

**Reply by danobi** [reply]
Currently stuck on https://sourceware.org/pipermail/elfutils-devel/2023q4/006759.html though :P. Maybe you know some libelf friends!

**Reply by jhuckaby** [reply]
Hmm, I'll ask around, but that is way above my head.  I can, however, get some more eyes on this project 👀.  I'm currently working on a blog post about my adventures in trying to write a "nethogs" style bandwidth monitor using bpftrace.  It works, with some caveats.  I hope to highlight your work on AOT compilation in the article, and hopefully get some people interested in helping.  🤞🏻

**Reply by danobi** [reply]
Went with shelling out to objcopy instead. Turns out section creation is a hard-ish problem. There's prior art with shelling out - pahole does this too. Can always be improved later.

**Reply by jhuckaby** [reply]
Glad you were able to figure something out!  I'm still working on my blog post 👍🏻
